rpcs3/Emu/Io/Dimensions.cpp

```diff
@@ -0,0 +1,620 @@
+	// the direction (0x00 in byte 6 for added) and their UID
+	std::array<u8, 32> figure_change_response = {0x56, 0x0b, figure.pad, 0x00, figure.index, 0x00};
+	std::memcpy(&figure_change_response[6], buf.data(), 7);
+	figure_change_response[13] = generate_checksum(figure_change_response, 13);
+	m_figure_added_removed_responses.push(figure_change_response);
+	return id;
+}
+
+bool dimensions_toypad::remove_figure(u8 pad, u8 index, bool save)
+{
+	std::lock_guard lock(dimensions_mutex);
+	dimensions_figure& figure = get_figure_by_index(index);
+	if (figure.index == 255)
+	{
+		return false;
+	}
+	// When a figure is removed from the toypad, respond to the game with the pad they were removed from, their index,
+	// the direction (0x01 in byte 6 for removed) and their UID
+	std::array<u8, 32> figure_change_response = {0x56, 0x0b, pad, 0x00, figure.index, 0x01};
+	std::memcpy(&figure_change_response[6], figure.data.data(), 7);
+	if (save)
+	{
+		figure.save();
+		figure.dim_file.close();
+	}
+	figure.index = 255;
+	figure.pad = 255;
+	figure_change_response[13] = generate_checksum(figure_change_response, 13);
+	m_figure_added_removed_responses.push(figure_change_response);
+	return true;
+}
+
+bool dimensions_toypad::move_figure(u8 pad, u8 index, u8 old_pad, u8 old_index)
+{
+	// When moving figures between spaces on the portal, remove any figure from the space they are moving to,
+	// then remove them from their current space, then load them to the space they are moving to
+	remove_figure(pad, index, true);
+
+	dimensions_figure& figure = get_figure_by_index(old_index);
+	const std::array<u8, 0x2D * 0x04> data = figure.data;
+	fs::file in_file = std::move(figure.dim_file);
+
+	remove_figure(old_pad, old_index, false);
+
+	load_figure(data, std::move(in_file), pad, index);
+
+	return true;
+}
+
+bool dimensions_toypad::create_blank_character(std::array<u8, 0x2D * 0x04>& buf, u16 id)
+{
+	random_uid(buf.data());
+	buf[7] = id & 0xFF;
+	std::array<u8, 7> uid = {buf[0], buf[1], buf[2], buf[4], buf[5], buf[6], buf[7]};
+
+	// Only characters are created with their ID encrypted and stored in pages 36 and 37,
+	// as well as a password stored in page 43. Blank tags have their information populated
+	// by the game when it calls the write_block command.
+	if (id != 0)
+	{
+		std::array<u8, 16> figure_key = generate_figure_key(buf);
+
+		std::array<u8, 8> value_to_encrypt = {};
+		write_to_ptr<le_t<u32>>(value_to_encrypt.data(), id);
+		write_to_ptr<le_t<u32>>(value_to_encrypt.data(), 4, id);
+
+		std::array<u8, 8> encrypted = encrypt(value_to_encrypt.data(), figure_key);
+
+		std::memcpy(&buf[36 * 4], &encrypted[0], 4);
+		std::memcpy(&buf[37 * 4], &encrypted[4], 4);
+
+		std::memcpy(&buf[43 * 4], pwd_generate(uid).data(), 4);
+	}
+	else
+	{
+		// Page 38 is used as verification for blank tags
+		write_to_ptr<be_t<u16>>(buf.data(), 38 * 4, 1);
+	}
+
+	return true;
+}
```

Comment:
write_to_ptr and read_to_ptr also take arrays as argument

```suggestion
bool dimensions_toypad::create_blank_character(std::array<u8, 0x2D * 0x04>& buf, u16 id)
{
	random_uid(buf.data());
	buf[7] = id & 0xFF;
	std::array<u8, 7> uid = {buf[0], buf[1], buf[2], buf[4], buf[5], buf[6], buf[7]};

	// Only characters are created with their ID encrypted and stored in pages 36 and 37,
	// as well as a password stored in page 43. Blank tags have their information populated
	// by the game when it calls the write_block command.
	if (id != 0)
	{
		const std::array<u8, 16> figure_key = generate_figure_key(buf);

		std::array<u8, 8> value_to_encrypt = {};
		write_to_ptr<le_t<u32>>(value_to_encrypt, id);
		write_to_ptr<le_t<u32>>(value_to_encrypt, 4, id);

		const std::array<u8, 8> encrypted = encrypt(value_to_encrypt.data(), figure_key);

		std::memcpy(&buf[36 * 4], &encrypted[0], 4);
		std::memcpy(&buf[37 * 4], &encrypted[4], 4);

		std::memcpy(&buf[43 * 4], pwd_generate(uid).data(), 4);
	}
	else
	{
		// Page 38 is used as verification for blank tags
		write_to_ptr<be_t<u16>>(buf, 38 * 4, 1);
	}

	return true;
}
```